src/SMPCoroutines.java

```diff
@@ -0,0 +1,217 @@
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.util.Scanner;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+/*
+* Inspired by
+* "Stable Marriage Problem by Coroutines"
+* Lloyd Allison - 1983
+* */
+public class SMPCoroutines {
+
+    class Man implements Runnable{
+        private int id;
+        private int[] preferences;
+        private int currentProposed;
+        private Woman[] womenList;
+
+        public Man(int id, int[] preferences) {
+            currentProposed = -1;
+            this.id = id;
+            this.preferences = preferences;
+        }
+
+        public void setWomenList(Woman[] womenList) {
+            this.womenList = womenList;
+        }
+
+        @Override
+        public void run() {
+            for (currentProposed = 0; currentProposed < preferences.length; currentProposed++) {
+                   proposeTo(preferences[currentProposed] -1);
+            }
+        }
+
+        private void proposeTo(int womanIndex) {
+            Woman next = womenList[womanIndex];
+            System.out.printf("Man %d proposed to %d\n", id, womanIndex);
+            next.receiveProposal(this.id);
+            try {
+                System.out.printf("Man %d waiting for response.\n", id);
+                synchronized (this) {
+                    this.wait();
+                }
+                System.out.printf("Man %d was rejected :(\n", id);
+            } catch (InterruptedException e) {
+
+            }
+        }
+
+    }
+
+
+    class Woman implements Runnable {
+        private int id;
+        private int[] preferences;
+
+        private int acceptedMan;
+        private Man[] menList;
+
+        private Semaphore sem;
+        private int proposedMan;
+
+        public Woman(int id, int[] preferences) {
+            this.id = id;
+            this.preferences = preferences;
+            acceptedMan = -1;
+            sem = new Semaphore(0);
+        }
+
+        public void setMenList(Man[] menList) {
+            this.menList = menList;
+        }
+
+        public void receiveProposal(int proposingMan) {
+            synchronized (sem) {
+                proposedMan = proposingMan;
+                sem.release();
+            }
+        }
+
+        @Override
+        public void run() {
+            while(true) {
```

Comment:
Maybe by modifying this condition is possible to get rid of the deadlock, it would involve checking if any of the men are still available or something similar